binder/src/main/java/io/grpc/binder/internal/BinderTransportSecurity.java

```diff
@@ -111,32 +154,20 @@ private static final class TransportAuthorizationState {
 
     /** Get whether we're authorized to make this call. */
     @CheckReturnValue
-    Status checkAuthorization(MethodDescriptor<?, ?> method) {
+    ListenableFuture<Status> checkAuthorization(MethodDescriptor<?, ?> method) {
       String serviceName = method.getServiceName();
       // Only cache decisions if the method can be sampled for tracing,
-      // which is true for all generated methods. Otherwise, programatically
-      // created methods could casue this cahe to grow unbounded.
+      // which is true for all generated methods. Otherwise, programmatically
+      // created methods could cause this cache to grow unbounded.
       boolean useCache = method.isSampledToLocalTracing();
-      Status authorization;
       if (useCache) {
-        authorization = serviceAuthorization.get(serviceName);
+        @Nullable ListenableFuture<Status> authorization = serviceAuthorization.get(serviceName);
         if (authorization != null) {
           return authorization;
         }
       }
-      try {
-        // TODO(10566): provide a synchronous version of "checkAuthorization" to avoid blocking the
-        // calling thread on the completion of the future.
-        authorization =
-            serverPolicyChecker.checkAuthorizationForServiceAsync(uid, serviceName).get();
-      } catch (ExecutionException e) {
-        // Do not cache this failure since it may be transient.
-        return Status.fromThrowable(e);
-      } catch (InterruptedException e) {
-        // Do not cache this failure since it may be transient.
-        Thread.currentThread().interrupt();
-        return Status.CANCELLED.withCause(e);
-      }
+      ListenableFuture<Status> authorization =
+          serverPolicyChecker.checkAuthorizationForServiceAsync(uid, serviceName);
```

Comment:
Suppose the async security policy is quite slow and a large number of requests that need checking come in while it is pending. How many calls to checkAuhtorizationAsync() do we want running concurrently? Should we just call it once and have all subsequent calls hinge on the same future?

Reply:
I think caching or deduplication would alter this class' API contract. For the original synchronous policies, the check would happen for every RPC (unless `useCache` was set). So clients had the freedom to implement non-constant checks, that return different results at different times for the same caller `uid` and target `serviceName`. Caching or deduplication for async policies would make them execute only on arbitrary occasions and may break developer's expectations.

Also, I don't have backing data, but ListenableFutures are meant to be highly scalable (thus why they are used in web servers with 100's+ of QPS). Clients are always welcome to cache the checks on their own (maybe we update the docs to infer so?).

Reply:
It might be OK to leave this out for now but I'm concerned because:
- useCache is almost always true (proto codegen'ed MethodDescriptors have setSampledToLocalTracing(true))
- Fixed-sized thread pools will eventually limit the number of concurrent synchronous SecurityPolicy checks in pathological cases today. With async SecurityPolicy these natural brakes are released.

I'm not concerned with the scalability of ListenableFutures themselves, it's the work happening behind the scenes that I feel should be limited to one operation outstanding per (method, uid).

Reply:
Re: "dynamic checks" - such a thing isn't really allowed according to the API contract for SecurityPolicy.checkAuthorization(). In particular: "As long as any given UID has active processes, this method should return the same value for that UID. In order words, policy changes which occur while a transport instance is active, will have no effect on that transport instance." This is what lets us cache the result of checkAuthorization() inside a single instance of BinderTransport.

Agreed that the current code might already be checking the same uid against the same SecurityPolicy today concurrently on two different threads. I believe this is harmless though due to the checkAuthorization() I quote above -- the answers must be the same.

How about leaving a todo here as a compromise for now?

Reply:
Got it. Added a comment stating this caveat and a TODO for #10669. Thanks for the insights!